idaes/apps/uncertainty_propagation/uncertainties.py

```diff
@@ -0,0 +1,371 @@
+    line_dic: dict
+        column numbers of the theta_names in the model. Index starts from 1
+    Raises
+    ------
+    RuntimeError
+        When ipopt or kaug or dotsens is not available
+    Exception
+        When ipopt fails 
+    """
+    #Create the solver plugin using the ASL interface
+    ipopt = SolverFactory('ipopt',solver_io='nl')
+    if solver_options is not None:
+        ipopt.options = solver_options
+    kaug = SolverFactory('k_aug',solver_io='nl')
+    dotsens = SolverFactory('dot_sens',solver_io='nl')
+    if not ipopt.available(False):
+        raise RuntimeError('ipopt is not available')
+    if not kaug.available(False):
+        raise RuntimeError('k_aug is not available')
+    if not dotsens.available(False):
+        raise RuntimeError('dotsens is not available')
+    
+    # Declare Suffixes
+    model.dual = Suffix(direction = Suffix.IMPORT)
+    model.ipopt_zL_out = Suffix(direction=Suffix.IMPORT)
+    model.ipopt_zU_out = Suffix(direction=Suffix.IMPORT)
+    model.ipopt_zL_in = Suffix(direction=Suffix.EXPORT)
+    model.ipopt_zU_in = Suffix(direction=Suffix.EXPORT)
+
+    # K_AUG SUFFIXES
+    model.dof_v = Suffix(direction=Suffix.EXPORT)  #: SUFFIX FOR K_AUG
+    model.rh_name = Suffix(direction=Suffix.IMPORT)  #: SUFFIX FOR K_AUG AS WELL
+    kaug.options["print_kkt"] = ""
+    results = ipopt.solve(model,tee=tee)
+
+    # Rasie Exception if ipopt fails 
+    if (results.solver.status == pyomo.opt.SolverStatus.warning):
+        raise Exception(results.solver.Message)
+    
+    for o in model.component_objects(Objective, active=True):
+        f_mean = value(o)
+    model.ipopt_zL_in.update(model.ipopt_zL_out)
+    model.ipopt_zU_in.update(model.ipopt_zU_out)
+    #: run k_aug
+    kaug.solve(model, tee=tee)  #: always call k_aug AFTER ipopt.
+    model.write('col_row.nl', format='nl', io_options={'symbolic_solver_labels':True})
+    # get the column numbers of theta
+    line_dic = {}
+    for v in theta_names:
+        line_dic[v] = line_num('col_row.col', v)
+    # load gradient of the objective function
+    gradient_f = np.loadtxt("./GJH/gradient_f_print.txt")
+    # load gradient of all constraints (sparse)
+    # If no constraint exists, return []
+    num_constraints = len(list(model.component_data_objects(Constraint,
+                                                            active=True,
+                                                            descend_into=True)))
+    if num_constraints > 0 :
+        gradient_c = np.loadtxt("./GJH/A_print.txt")
+    else:
+        gradient_c = np.array([])
+
+    # remove all generated files
+    shutil.move("col_row.nl", "./GJH/")
+    shutil.move("col_row.col", "./GJH/")
+    shutil.move("col_row.row", "./GJH/")
+    shutil.rmtree('GJH', ignore_errors=True)
+    return gradient_f, gradient_c, line_dic
+
+def line_num(file_name, target):
+    """This function returns the line number contains 'target' in the file_name
```

Comment:
Why is this function important? Please explain where it is used in the docstring.

Reply:
This function finds constraints (rows) that include uncertain variables. I added comments.